### PR TITLE
Add sort verification for writer fuzzer

### PR DESCRIPTION
**Pull request by kewang1024**

### CODE REVIEW

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 1]
```diff
@@ -162,6 +168,39 @@ class WriterFuzzer {
     return partitionCount > 0 && bucketCount > 0;
   }
 
+  // Check the table properties and see if the table is sorted.
+  bool isSorted(
+      const int32_t partitionCount,
+      const int32_t bucketCount,
+      const int32_t sortColumnCount) {
+    return partitionCount > 0 && bucketCount > 0 && sortColumnCount > 0;
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:176]: Shall we just check on sortColumnCount? Why we need to check partitionCount and bucketCount? We can put an assertion either in Prestissimo or Table writer that sort columns shouldn't be set if it is not partition and bucket table? thanks!

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 2]
```diff
@@ -162,6 +168,39 @@ class WriterFuzzer {
...
+  // For example, for a table sorted by age, reading a split that belongs to ds
+  // = 2022-01-01 and bucket 1:
+  // SELECT age FROM temp_write where ds = '2022-01-01' and "$bucket" = 1
+  std::string sortSql(
+      const std::shared_ptr<HiveConnectorSplit>& split,
+      const std::vector<std::string>& names,
+      const std::vector<TypePtr>& types,
+      int32_t partitionOffset,
+      const std::vector<std::string>& partitionKeys,
+      std::vector<std::shared_ptr<const HiveSortingColumn>> sortBy);
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:190]:
```
const std::vector<std::shared_ptr<const HiveSortingColumn>>&
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 3]
```diff
@@ -162,6 +168,39 @@ class WriterFuzzer {
...
+  // When concatenating a partition value, if it's non-varchar, no change, eg:
+  // age = 10
+  // If it's varchar, we need to add single quote and also escape the single
+  // quote in partition value, eg:
+  // city = '''SF'''
+  std::string partitionToSql(
+      const TypePtr& type,
+      const std::string& partitionValue);
+
+  // Check if two vectors have the same data including their orders.
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:201]:
```
s/Check/Checks/
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 4]
```diff
@@ -259,10 +319,27 @@ void WriterFuzzer::go() {
...
     std::vector<std::string> bucketColumns;
     std::vector<std::string> partitionKeys;
+    std::vector<std::shared_ptr<const HiveSortingColumn>> sortBy;
 
     // Regular table columns
     generateColumns(5, "c", kRegularColumnTypes_, 2, names, types);
 
+    // TODO: sort columns can overlap as bucket columns
+    // 50% of times test ordered write.
+    const auto sortColumnOffset = names.size();
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:329]: Shall we only set this after we decide the bucket and partition columns? thanks!

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 5]
```diff
@@ -596,5 +717,66 @@ RowTypePtr WriterFuzzer::generateOutputType(
   return {ROW(std::move(outputNames), std::move(outputTypes))};
 }
 
+std::string WriterFuzzer::sortSql(
+    const std::shared_ptr<HiveConnectorSplit>& split,
+    const std::vector<std::string>& names,
+    const std::vector<TypePtr>& types,
+    int32_t partitionOffset,
+    const std::vector<std::string>& partitionKeys,
+    const std::vector<std::shared_ptr<const HiveSortingColumn>> sortBy) {
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:726]:
```
const std::vector<std::shared_ptr<const HiveSortingColumn>>&
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 6]
```diff
@@ -596,5 +717,66 @@ RowTypePtr WriterFuzzer::generateOutputType(
...
+    const TypePtr& type,
+    const std::string& partitionValue) {
+  if (type->isVarchar()) {
+    return "'" + std::regex_replace(partitionValue, std::regex("'"), "''") +
+        "'";
+  }
+  return partitionValue;
+}
+
+void WriterFuzzer::checkEqualVectors(
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:765]: Can we use assertEqualResults? in QueryAsserations.h? Thanks!

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 7]
```diff
@@ -422,6 +506,43 @@ void WriterFuzzer::verifyWriter(
...
+    // Read from each file and check if data is sorted as presto sorted result.
+    for (const auto& split : splits) {
+      auto readSingleSplitPlan =
+          PlanBuilder()
+              .tableScan(
+                  generateOutputType(sortColumnNames, sortColumnTypes, 0, 0))
+              .planNode();
+      auto singleSplitData = execute(readSingleSplitPlan, 1, {split});
+
+      auto singleSplitSql = sortSql(
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:527]: Shall we move sortSql to reference DB API? thanks!

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 8]
```diff
@@ -422,6 +506,43 @@ void WriterFuzzer::verifyWriter(
...
+    const std::vector<std::string> sortColumnNames = {
+        names.begin() + sortColumnOffset,
+        names.begin() + sortColumnOffset + sortBy.size()};
+    const std::vector<TypePtr> sortColumnTypes = {
+        types.begin() + sortColumnOffset,
+        types.begin() + sortColumnOffset + sortBy.size()};
+
+    // Read from each file and check if data is sorted as presto sorted result.
+    for (const auto& split : splits) {
+      auto readSingleSplitPlan =
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:520]:
```
s/readSingleSplitPlan/splitReadPlan or just readPlan/
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 9]
```diff
@@ -422,6 +506,43 @@ void WriterFuzzer::verifyWriter(
...
+      auto singleSplitSql = sortSql(
+          std::dynamic_pointer_cast<HiveConnectorSplit>(split.connectorSplit),
+          names,
+          types,
+          partitionOffset,
+          partitionKeys,
+          sortBy);
+
+      auto dataVector = referenceQueryRunner_->execute(singleSplitSql);
+      auto reference_data = dataVector.at(0);
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:536]:
```
s/reference_data/referenceData/
```

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 10]
```diff
@@ -422,6 +506,43 @@ void WriterFuzzer::verifyWriter(
...
+
+      auto singleSplitSql = sortSql(
+          std::dynamic_pointer_cast<HiveConnectorSplit>(split.connectorSplit),
+          names,
+          types,
+          partitionOffset,
+          partitionKeys,
+          sortBy);
+
+      auto dataVector = referenceQueryRunner_->execute(singleSplitSql);
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:535]:
```
s/dataVector/const auto referenceResult/
```

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 thanks % minors.

File: velox/exec/fuzzer/WriterFuzzer.cpp (hunk 6 above)
Comment (kewang1024) [reply velox/exec/fuzzer/WriterFuzzer.cpp:765]: That will compare result without considering order

File: velox/exec/fuzzer/WriterFuzzer.cpp (hunk 7 above)
Comment (kewang1024) [reply velox/exec/fuzzer/WriterFuzzer.cpp:507]: Most of the logic is more about converting split information to sql, which is not applicable to move to referenceDB API, so I will keep it here

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 11]
```diff
@@ -579,20 +687,64 @@ std::map<std::string, int32_t> WriterFuzzer::getPartitionNameAndFilecount(
...
 }
 
+std::string WriterFuzzer::sortSql(
+    const std::shared_ptr<HiveConnectorSplit>& split,
+    const std::vector<std::string>& names,
+    const std::vector<TypePtr>& types,
+    int32_t partitionOffset,
+    const std::vector<std::string>& partitionKeys,
+    const std::vector<std::shared_ptr<const HiveSortingColumn>>& sortBy) {
+  // For a split, extract the partition filters and bucket filters.
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:711]: Can we parse the split info and then pass to reference DB? So we don't need to expose this split directly to reference DB? thanks!

**Review by xiaoxmeng — COMMENTED** [review]
Can you add PR description before land?

File: velox/exec/fuzzer/WriterFuzzer.cpp [hunk 12]
```diff
@@ -403,23 +474,60 @@ void WriterFuzzer::verifyWriter(
...
+
+    // Read from each file and check if data is sorted as presto sorted result.
+    for (const auto& split : splits) {
+      auto splitReadPlan = PlanBuilder()
+                               .tableScan(generateOutputType(
+                                   sortColumnNames, sortColumnTypes, 0))
+                               .planNode();
+      auto singleSplitData = execute(splitReadPlan, 1, {split});
+
+      auto singleSplitSql = sortSql(
```
Comment (xiaoxmeng) [inline velox/exec/fuzzer/WriterFuzzer.cpp:510]:
```
s/singleSplitSql/const singleSplitReferenceSql/
```

**Review by xiaoxmeng — APPROVED** [review]
@kewang1024 thanks!

**Pull request closed without merging**